bridge-fungible-token/src/utils.sw

```diff
@@ -1,11 +1,187 @@
+        && u64s.3 >= 1_000_000_000
+        && u64s.0 == 0
+        && u64s.1 == 0
+        && u64s.2 == 0
+    {
+        // reduce decimals by 9 places
+        Result::Ok(u64s.3 / 1_000_000_000)
+    } else {
+        Result::Err(BridgeFungibleTokenError::BridgedValueIncompatability)
+    }
+}
+
+/// Build a single b256 value from a u64 left-padded with 3 0u64's
+pub fn u64_to_b256(val: u64) -> b256 {
+    let res: b256 = 0x0000000000000000000000000000000000000000000000000000000000000000;
+    let ptr = addr_of(res);
+    write(ptr, 0);
+    write(ptr + 8, 0);
+    write(ptr + 16, 0);
+    write(ptr + 24, val);
+    res
+}
+
+/// Get 4 64 bit words from a single b256 value.
+pub fn decompose(val: b256) -> (u64, u64, u64, u64) {
+    let w1 = single_word_from_b256(val, 0);
+    let w2 = single_word_from_b256(val, 8);
+    let w3 = single_word_from_b256(val, 16);
+    let w4 = single_word_from_b256(val, 24);
+    (w1, w2, w3, w4)
+}
+
+/// Extract a single 64 bit word from a b256 value using the specified offset.
+fn single_word_from_b256(val: b256, offset: u64) -> u64 {
+    let mut empty: u64 = 0;
+    asm(r1: val, offset: offset, r2, res: empty) {
+        add r2 r1 offset;
+        lw res r2 i0;
+        res: u64
+    }
+}
+
+pub fn parse_message_data(msg_idx: u8) -> MessageData {
+    let mut msg_data = MessageData {
+        fuel_token: ~ContractId::from(ZERO_B256),
+        l1_asset: ~EvmAddress::from(ZERO_B256),
+        from: ~EvmAddress::from(ZERO_B256),
+        to: ~Address::from(ZERO_B256),
+        amount: ZERO_B256,
+    };
+
+    // Parse the message data
+    msg_data.fuel_token = ~ContractId::from(input_message_data::<b256>(msg_idx, 0));
+    msg_data.l1_asset = ~EvmAddress::from(input_message_data::<b256>(msg_idx, 32));
+    msg_data.from = ~EvmAddress::from(input_message_data::<b256>(msg_idx, 32 + 32));
+    msg_data.to = ~Address::from(input_message_data::<b256>(msg_idx, 32 + 32 + 32));
+    msg_data.amount = input_message_data::<b256>(msg_idx, 32 + 32 + 32 + 32);
+
+    msg_data
+}
+pub fn encode_data(to: b256, amount: u64) -> Vec<u64> {
+    let mut data = ~Vec::with_capacity(13);
+    // start with the function selector for finalizeWithdrawal on the L1ERC20Gateway contract
+    data.push(0x53ef1461);
+
+    // add the address to recieve coins
+    let (recip_1, recip_2, recip_3, recip_4) = decompose(to);
+    data.push(recip_1);
+    data.push(recip_2);
+    data.push(recip_3);
+    data.push(recip_4);
+
+    // add the address of the L1 token contract
+    let (token_1, token_2, token_3, token_4) = decompose(LAYER_1_TOKEN);
+    data.push(token_1);
+    data.push(token_2);
+    data.push(token_3);
+    data.push(token_4);
+
+    // add the amount of tokens, padding with 3 0u64s to allow reading as a b256 on the other side of the bridge.
+    data.push(0u64);
+    data.push(0u64);
+    data.push(0u64);
+    data.push(amount);
+
+    data
+}
+
+pub fn send_message_output(to: b256, amount: u64, ) {
```

Comment:
`amount` should be a `b256`. Any conversion of decimals should happen outside of this function.